### PR TITLE
Support reusing manifest content across multiple projects

### PR DESCRIPTION
**Pull request by pepibumur**
### Short description 📝
The goal of this PR is to support reusing manifest content across project by extracting it into Swift files that are bundled into a module that the manifest is liked against.

### Implementation 👩‍💻👨‍💻
- [x] Add a new manifest loader that compiles the module in a temporary directory.
- [x] Adjust the manifest loader to link against that module if it's created.
- [x] Adjust the project generation logic to generate a new target that includes those files.
- [x] Investigate why the output from `swiftc` is not printed when the command fails.
- [x] Add unit tests
- [x] Add acceptance tests.

### Example

The example below shows how a project could leverage this feature to define a function that acts as a factory of projects:

image

### CODE REVIEW

**Comment by ollieatkinson** [comment]
Would we want to support projects with multiple Tuist directories? A good example of this would be if you include a Carthage dependency which is also a Tuist project.

I like the direction this is taking as it will open up the possibility of creating a specific workspace for editing the helpers & manifests (auto-completion and all the normal help from the swift compiler).

**Comment by pepibumur** [comment]
So you'd expect the helpers to be in the repositories that Carthage clones?

**Comment by ollieatkinson** [comment]
> So you'd expect the helpers to be in the repositories that Carthage clones?

Not just in the carthage repositories, but specifically that each repository may contain it's own helpers;

**Comment by pepibumur** [comment]
> Not just in the carthage repositories, but specifically that each repository may contain it's own helpers;

How would each of those projects use the helpers? Sorry for asking so many question but I'd like to understand the need very well to see if we can come up with something that covers your scenario too.

File: Sources/TuistKit/Utils/RootDirectoryLocator.swift [hunk 1]
```diff
@@ -21,7 +21,7 @@ final class RootDirectoryLocator: RootDirectoryLocating {
     func locate(from path: AbsolutePath) -> AbsolutePath? {
         if let cachedDirectory = cached(path: path) {
             return cachedDirectory
-        } else if let tuistDirectory = FileHandler.shared.locateDirectory(Constants.tuistFolderName, traversingFrom: path) {
+        } else if let tuistDirectory = FileHandler.shared.locateDirectory(Constants.tuistDirectoryName, traversingFrom: path) {
```
Comment (kwridan) [inline Sources/TuistKit/Utils/RootDirectoryLocator.swift]: @ollieatkinson the issue you flagged does indeed occur (multiple Tuist directories). 

To mitigate this we'd need to handle the caching differently, the [`cached`](https://github.com/tuist/tuist/pull/542/files#diff-0c124dde2b83e5dd1ee0917ecfa27039R39) method would need to only return a cached path if one is present without traversing the cache.

i.e.

- Check `path` in cache
- Check the file system
- Repeat for the parent

Comment (kwridan) [reply Sources/TuistKit/Utils/RootDirectoryLocator.swift]: #630 should fix this

File: Tests/TuistGeneratorTests/Generator/WorkspaceStructureGeneratorTests.swift [hunk 2]
```diff
@@ -329,6 +329,10 @@ class WorkspaceStructureGeneratorTests: XCTestCase {
             return cache[path] == .folder
         }
 
+        func locateDirectoryTraversingParents(from _: AbsolutePath, path _: String) -> AbsolutePath? {
```
Comment (kwridan) [inline Tests/TuistGeneratorTests/Generator/WorkspaceStructureGeneratorTests.swift]: might be a redundant method?

**Comment by pepibumur** [comment]
This work is not user-ready yet. I [opened an issue](https://github.com/tuist/tuist/issues/668) to continue the work on follow-up PRs.

**Pull request merged**